File: python/classes/bank_account.py
"""Bank Account Manager — double-entry ledger, atomic transactions, balance as aggregate.

A transaction is a list of entries, each a signed amount against one account.
Transfers have two entries whose amounts sum to zero (conservation). Deposits
and withdrawals have one entry — money crossing the bank's boundary. Balance
is not stored; it is the signed sum of all entries for an account (same
event-sourcing pattern as G048, now with multi-sided events).

Money is integer cents throughout. No float arithmetic on money, ever.
"""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Optional
import itertools


class AccountKind(Enum):
    CHECKING = "checking"
    SAVINGS = "savings"
    CREDIT = "credit"


class AccountStatus(Enum):
    OPEN = "open"
    FROZEN = "frozen"
    CLOSED = "closed"


class TxKind(Enum):
    DEPOSIT = "deposit"
    WITHDRAWAL = "withdrawal"
    TRANSFER = "transfer"
    FEE = "fee"
    INTEREST = "interest"


@dataclass
class Customer:
    id: str
    name: str


@dataclass
class Account:
    id: str
    holder_id: str
    kind: AccountKind
    status: AccountStatus = AccountStatus.OPEN
    opened_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


@dataclass
class Entry:
    """One side of a transaction. Positive credits the account, negative debits it."""
    account_id: str
    amount_cents: int


@dataclass
class Transaction:
    id: int
    kind: TxKind
    entries: list[Entry]
    memo: str = ""
    at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


class BankError(Exception):
    """Base for bank errors."""


class UnknownEntity(BankError): ...
class InsufficientFunds(BankError): ...
class AccountFrozen(BankError): ...
class AccountClosed(BankError): ...
class NonZeroBalance(BankError): ...
class UnbalancedTransfer(BankError): ...


class Bank:
    """Ledger-first bank. Balances are projections over the transaction log."""

    def __init__(self) -> None:
        self.customers: dict[str, Customer] = {}
        self.accounts: dict[str, Account] = {}
        self.transactions: dict[int, Transaction] = {}
        self._tx_ids = itertools.count(1)
        self._acct_ids = itertools.count(1)

    # --- registration ---
    def add_customer(self, customer: Customer) -> None:
        if customer.id in self.customers:
            raise ValueError(f"customer already registered: {customer.id}")
        self.customers[customer.id] = customer

    def open_account(self, holder_id: str, kind: AccountKind,
                     account_id: Optional[str] = None) -> Account:
        if holder_id not in self.customers:
            raise UnknownEntity(f"unknown customer: {holder_id}")
        aid = account_id or f"A{next(self._acct_ids):06d}"
        if aid in self.accounts:
            raise ValueError(f"account id already in use: {aid}")
        account = Account(id=aid, holder_id=holder_id, kind=kind)
        self.accounts[aid] = account
        return account

    def close_account(self, account_id: str) -> Account:
        account = self._require_account(account_id)
        if self.balance(account_id) != 0:
            raise NonZeroBalance(
                f"cannot close {account_id} with balance {self.balance(account_id)} cents"
            )
        account.status = AccountStatus.CLOSED
        return account

    def freeze(self, account_id: str) -> Account:
        account = self._require_account(account_id)
        account.status = AccountStatus.FROZEN
        return account

    def unfreeze(self, account_id: str) -> Account:
        account = self._require_account(account_id)
        if account.status == AccountStatus.FROZEN:
            account.status = AccountStatus.OPEN
        return account

    def _require_account(self, account_id: str) -> Account:
        if account_id not in self.accounts:
            raise UnknownEntity(f"unknown account: {account_id}")
        return self.accounts[account_id]

    def _require_active(self, account_id: str) -> Account:
        account = self._require_account(account_id)
        if account.status == AccountStatus.FROZEN:
            raise AccountFrozen(f"account frozen: {account_id}")
        if account.status == AccountStatus.CLOSED:
            raise AccountClosed(f"account closed: {account_id}")
        return account

    # --- transactions ---
    def deposit(self, account_id: str, amount_cents: int, memo: str = "") -> Transaction:
        if amount_cents <= 0:
            raise ValueError("deposit amount must be positive cents")
        self._require_active(account_id)
        return self._commit(TxKind.DEPOSIT, [Entry(account_id, amount_cents)], memo)

    def withdraw(self, account_id: str, amount_cents: int, memo: str = "") -> Transaction:
        if amount_cents <= 0:
            raise ValueError("withdraw amount must be positive cents")
        self._require_active(account_id)
        if self.balance(account_id) < amount_cents:
            raise InsufficientFunds(
                f"{account_id}: balance {self.balance(account_id)}, requested {amount_cents}"
            )
        return self._commit(TxKind.WITHDRAWAL, [Entry(account_id, -amount_cents)], memo)

    def transfer(self, from_id: str, to_id: str, amount_cents: int,
                 memo: str = "") -> Transaction:
        if amount_cents <= 0:
            raise ValueError("transfer amount must be positive cents")
        if from_id == to_id:
            raise ValueError("transfer requires distinct accounts")
        self._require_active(from_id)
        self._require_active(to_id)
        if self.balance(from_id) < amount_cents:
            raise InsufficientFunds(
                f"{from_id}: balance {self.balance(from_id)}, requested {amount_cents}"
            )
        entries = [Entry(from_id, -amount_cents), Entry(to_id, amount_cents)]
        if sum(e.amount_cents for e in entries) != 0:
            raise UnbalancedTransfer("transfer entries must sum to zero")
        return self._commit(TxKind.TRANSFER, entries, memo)

    def _commit(self, kind: TxKind, entries: list[Entry], memo: str) -> Transaction:
        tx = Transaction(id=next(self._tx_ids), kind=kind, entries=entries, memo=memo)
        self.transactions[tx.id] = tx
        return tx

    # --- aggregates ---
    def balance(self, account_id: str) -> int:
        self._require_account(account_id)
        total = 0
        for tx in self.transactions.values():
            for e in tx.entries:
                if e.account_id == account_id:
                    total += e.amount_cents
        return total

    def statement(self, account_id: str,
                  from_time: Optional[datetime] = None,
                  to_time: Optional[datetime] = None) -> list[Transaction]:
        self._require_account(account_id)
        result: list[Transaction] = []
        for tx in self.transactions.values():
            if not any(e.account_id == account_id for e in tx.entries):
                continue
            if from_time is not None and tx.at < from_time:
                continue
            if to_time is not None and tx.at > to_time:
                continue
            result.append(tx)
        return sorted(result, key=lambda t: t.at)

    def total_assets(self, customer_id: str) -> int:
        if customer_id not in self.customers:
            raise UnknownEntity(f"unknown customer: {customer_id}")
        return sum(self.balance(a.id) for a in self.accounts.values()
                   if a.holder_id == customer_id)

    def ledger_is_balanced(self) -> bool:
        """Global invariant: the sum of all transfer entries is zero.

        Deposits and withdrawals cross the bank boundary and do not balance
        internally; transfers must. This check isolates transfers.
        """
        total = 0
        for tx in self.transactions.values():
            if tx.kind == TxKind.TRANSFER:
                total += sum(e.amount_cents for e in tx.entries)
        return total == 0


# --- demo ---
if __name__ == "__main__":
    bank = Bank()
    bank.add_customer(Customer("C001", "Nathan"))
    bank.add_customer(Customer("C002", "Korrallan"))

    checking = bank.open_account("C001", AccountKind.CHECKING, account_id="ACC-CHK-1")
    savings  = bank.open_account("C001", AccountKind.SAVINGS,  account_id="ACC-SAV-1")
    joint    = bank.open_account("C002", AccountKind.CHECKING, account_id="ACC-CHK-2")

    bank.deposit(checking.id, 500_00, memo="opening deposit")
    bank.transfer(checking.id, savings.id, 200_00, memo="monthly savings")
    bank.transfer(checking.id, joint.id,   75_00,  memo="groceries split")

    try:
        bank.withdraw(checking.id, 1_000_00, memo="big ATM")
    except InsufficientFunds as e:
        print(f"refused: {e}")

    print(f"checking: {bank.balance(checking.id)} cents")
    print(f"savings:  {bank.balance(savings.id)}  cents")
    print(f"joint:    {bank.balance(joint.id)}    cents")
    print(f"Nathan total assets: {bank.total_assets('C001')} cents")
    print(f"ledger balanced (transfers sum to zero): {bank.ledger_is_balanced()}")

File: rust/classes/bank_account.rs
//! Bank Account Manager — double-entry ledger, atomic transactions, balance as aggregate.
//!
//! A transaction is a list of entries, each a signed amount against one account.
//! Transfers sum to zero (conservation). Deposits and withdrawals cross the bank
//! boundary as single entries. Balance is the signed sum of an account's entries.
//! Money is integer cents throughout.

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AccountKind {
    Checking,
    Savings,
    Credit,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AccountStatus {
    Open,
    Frozen,
    Closed,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TxKind {
    Deposit,
    Withdrawal,
    Transfer,
    Fee,
    Interest,
}

#[derive(Debug, Clone)]
pub struct Customer {
    pub id: String,
    pub name: String,
}

#[derive(Debug, Clone)]
pub struct Account {
    pub id: String,
    pub holder_id: String,
    pub kind: AccountKind,
    pub status: AccountStatus,
}

#[derive(Debug, Clone)]
pub struct Entry {
    pub account_id: String,
    pub amount_cents: i64,
}

#[derive(Debug, Clone)]
pub struct Transaction {
    pub id: u64,
    pub kind: TxKind,
    pub entries: Vec<Entry>,
    pub memo: String,
}

#[derive(Debug)]
pub enum BankError {
    UnknownCustomer(String),
    UnknownAccount(String),
    Duplicate(String),
    InvalidAmount,
    InsufficientFunds { account_id: String, balance: i64, requested: i64 },
    AccountFrozen(String),
    AccountClosed(String),
    NonZeroBalance { account_id: String, balance: i64 },
    UnbalancedTransfer,
    SameAccountTransfer,
}

pub struct Bank {
    pub customers: HashMap<String, Customer>,
    pub accounts: HashMap<String, Account>,
    pub transactions: HashMap<u64, Transaction>,
    next_tx_id: u64,
    next_account_seq: u64,
}

impl Bank {
    pub fn new() -> Self {
        Self {
            customers: HashMap::new(),
            accounts: HashMap::new(),
            transactions: HashMap::new(),
            next_tx_id: 1,
            next_account_seq: 1,
        }
    }

    pub fn add_customer(&mut self, customer: Customer) -> Result<(), BankError> {
        if self.customers.contains_key(&customer.id) {
            return Err(BankError::Duplicate(customer.id));
        }
        self.customers.insert(customer.id.clone(), customer);
        Ok(())
    }

    pub fn open_account(
        &mut self,
        holder_id: &str,
        kind: AccountKind,
        account_id: Option<&str>,
    ) -> Result<String, BankError> {
        if !self.customers.contains_key(holder_id) {
            return Err(BankError::UnknownCustomer(holder_id.to_string()));
        }
        let id = match account_id {
            Some(s) => s.to_string(),
            None => {
                let n = self.next_account_seq;
                self.next_account_seq += 1;
                format!("A{:06}", n)
            }
        };
        if self.accounts.contains_key(&id) {
            return Err(BankError::Duplicate(id));
        }
        self.accounts.insert(
            id.clone(),
            Account {
                id: id.clone(),
                holder_id: holder_id.to_string(),
                kind,
                status: AccountStatus::Open,
            },
        );
        Ok(id)
    }

    pub fn close_account(&mut self, account_id: &str) -> Result<(), BankError> {
        let balance = self.balance(account_id)?;
        if balance != 0 {
            return Err(BankError::NonZeroBalance {
                account_id: account_id.to_string(),
                balance,
            });
        }
        let account = self
            .accounts
            .get_mut(account_id)
            .ok_or_else(|| BankError::UnknownAccount(account_id.to_string()))?;
        account.status = AccountStatus::Closed;
        Ok(())
    }

    pub fn freeze(&mut self, account_id: &str) -> Result<(), BankError> {
        let account = self
            .accounts
            .get_mut(account_id)
            .ok_or_else(|| BankError::UnknownAccount(account_id.to_string()))?;
        account.status = AccountStatus::Frozen;
        Ok(())
    }

    pub fn unfreeze(&mut self, account_id: &str) -> Result<(), BankError> {
        let account = self
            .accounts
            .get_mut(account_id)
            .ok_or_else(|| BankError::UnknownAccount(account_id.to_string()))?;
        if account.status == AccountStatus::Frozen {
            account.status = AccountStatus::Open;
        }
        Ok(())
    }

    fn require_active(&self, account_id: &str) -> Result<(), BankError> {
        let account = self
            .accounts
            .get(account_id)
            .ok_or_else(|| BankError::UnknownAccount(account_id.to_string()))?;
        match account.status {
            AccountStatus::Open => Ok(()),
            AccountStatus::Frozen => Err(BankError::AccountFrozen(account_id.to_string())),
            AccountStatus::Closed => Err(BankError::AccountClosed(account_id.to_string())),
        }
    }

    pub fn deposit(&mut self, account_id: &str, amount_cents: i64, memo: &str) -> Result<u64, BankError> {
        if amount_cents <= 0 {
            return Err(BankError::InvalidAmount);
        }
        self.require_active(account_id)?;
        Ok(self.commit(
            TxKind::Deposit,
            vec![Entry { account_id: account_id.to_string(), amount_cents }],
            memo,
        ))
    }

    pub fn withdraw(&mut self, account_id: &str, amount_cents: i64, memo: &str) -> Result<u64, BankError> {
        if amount_cents <= 0 {
            return Err(BankError::InvalidAmount);
        }
        self.require_active(account_id)?;
        let balance = self.balance(account_id)?;
        if balance < amount_cents {
            return Err(BankError::InsufficientFunds {
                account_id: account_id.to_string(),
                balance,
                requested: amount_cents,
            });
        }
        Ok(self.commit(
            TxKind::Withdrawal,
            vec![Entry { account_id: account_id.to_string(), amount_cents: -amount_cents }],
            memo,
        ))
    }

    pub fn transfer(
        &mut self,
        from_id: &str,
        to_id: &str,
        amount_cents: i64,
        memo: &str,
    ) -> Result<u64, BankError> {
        if amount_cents <= 0 {
            return Err(BankError::InvalidAmount);
        }
        if from_id == to_id {
            return Err(BankError::SameAccountTransfer);
        }
        self.require_active(from_id)?;
        self.require_active(to_id)?;
        let balance = self.balance(from_id)?;
        if balance < amount_cents {
            return Err(BankError::InsufficientFunds {
                account_id: from_id.to_string(),
                balance,
                requested: amount_cents,
            });
        }
        let entries = vec![
            Entry { account_id: from_id.to_string(), amount_cents: -amount_cents },
            Entry { account_id: to_id.to_string(), amount_cents },
        ];
        let sum: i64 = entries.iter().map(|e| e.amount_cents).sum();
        if sum != 0 {
            return Err(BankError::UnbalancedTransfer);
        }
        Ok(self.commit(TxKind::Transfer, entries, memo))
    }

    fn commit(&mut self, kind: TxKind, entries: Vec<Entry>, memo: &str) -> u64 {
        let id = self.next_tx_id;
        self.next_tx_id += 1;
        self.transactions.insert(
            id,
            Transaction { id, kind, entries, memo: memo.to_string() },
        );
        id
    }

    pub fn balance(&self, account_id: &str) -> Result<i64, BankError> {
        if !self.accounts.contains_key(account_id) {
            return Err(BankError::UnknownAccount(account_id.to_string()));
        }
        Ok(self
            .transactions
            .values()
            .flat_map(|t| t.entries.iter())
            .filter(|e| e.account_id == account_id)
            .map(|e| e.amount_cents)
            .sum())
    }

    pub fn total_assets(&self, customer_id: &str) -> Result<i64, BankError> {
        if !self.customers.contains_key(customer_id) {
            return Err(BankError::UnknownCustomer(customer_id.to_string()));
        }
        let mut total = 0;
        for a in self.accounts.values() {
            if a.holder_id == customer_id {
                total += self.balance(&a.id)?;
            }
        }
        Ok(total)
    }

    /// Global invariant: all transfer transactions must sum to zero.
    pub fn ledger_is_balanced(&self) -> bool {
        self.transactions
            .values()
            .filter(|t| t.kind == TxKind::Transfer)
            .all(|t| t.entries.iter().map(|e| e.amount_cents).sum::<i64>() == 0)
    }
}

impl Default for Bank {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn setup() -> (Bank, String, String, String) {
        let mut b = Bank::new();
        b.add_customer(Customer { id: "C1".into(), name: "A".into() }).unwrap();
        b.add_customer(Customer { id: "C2".into(), name: "B".into() }).unwrap();
        let a1 = b.open_account("C1", AccountKind::Checking, Some("A1")).unwrap();
        let a2 = b.open_account("C1", AccountKind::Savings, Some("A2")).unwrap();
        let a3 = b.open_account("C2", AccountKind::Checking, Some("A3")).unwrap();
        (b, a1, a2, a3)
    }

    #[test]
    fn deposit_increases_balance() {
        let (mut b, a1, _, _) = setup();
        b.deposit(&a1, 10_000, "").unwrap();
        assert_eq!(b.balance(&a1).unwrap(), 10_000);
    }

    #[test]
    fn withdraw_reduces_balance() {
        let (mut b, a1, _, _) = setup();
        b.deposit(&a1, 10_000, "").unwrap();
        b.withdraw(&a1, 3_000, "").unwrap();
        assert_eq!(b.balance(&a1).unwrap(), 7_000);
    }

    #[test]
    fn withdraw_refuses_overdraft() {
        let (mut b, a1, _, _) = setup();
        b.deposit(&a1, 1_000, "").unwrap();
        assert!(matches!(
            b.withdraw(&a1, 2_000, ""),
            Err(BankError::InsufficientFunds { .. })
        ));
    }

    #[test]
    fn transfer_conserves_total() {
        let (mut b, a1, a2, _) = setup();
        b.deposit(&a1, 10_000, "").unwrap();
        b.transfer(&a1, &a2, 4_000, "").unwrap();
        assert_eq!(b.balance(&a1).unwrap(), 6_000);
        assert_eq!(b.balance(&a2).unwrap(), 4_000);
        assert!(b.ledger_is_balanced());
    }

    #[test]
    fn frozen_account_rejects_transactions() {
        let (mut b, a1, _, _) = setup();
        b.deposit(&a1, 1_000, "").unwrap();
        b.freeze(&a1).unwrap();
        assert!(matches!(b.deposit(&a1, 100, ""), Err(BankError::AccountFrozen(_))));
        assert!(matches!(b.withdraw(&a1, 100, ""), Err(BankError::AccountFrozen(_))));
    }

    #[test]
    fn close_requires_zero_balance() {
        let (mut b, a1, _, _) = setup();
        b.deposit(&a1, 100, "").unwrap();
        assert!(matches!(
            b.close_account(&a1),
            Err(BankError::NonZeroBalance { .. })
        ));
        b.withdraw(&a1, 100, "").unwrap();
        b.close_account(&a1).unwrap();
        assert!(matches!(b.deposit(&a1, 100, ""), Err(BankError::AccountClosed(_))));
    }
}
